blog/2024-11-mastering-effect-and-afterRenderEffect/README.md

```diff
@@ -0,0 +1,477 @@
+        console.warn(`read executes`);
+        console.log('read: new height:', newHeight());
+      }
+    });
+
+    // Trigger a new run every 4 seconds by setting the signal `extraHeight`
+    setInterval(() => {
+      console.warn('---- new round ----');
+      this.extraHeight.update(x => ++x)
+    }, 4_000);
+
+    // Try this, if the signal value stays the same, nothing will hapen
+    // setInterval(() => this.extraHeight.update(x => x), 4_000);
+
+    // cleanup callbacks are also executed when we destroy the hook
+    // setTimeout(() => effect.destroy(), 20_000);
+  }
+}
+```
+
+In our setup, an interval updates the `extraHeight` signal every 4 seconds.
+By updating the signal `extraHeight`, we create a "dirty" state that restarts the `afterRenderEffect()` phases, which checks and adjusts the height of the `<textarea>` as needed:
+
+**Explanation of the Phases**  
+
+In this example, an interval updates `extraHeight` every 4 seconds, creating a new round of execution across the phases. 
+Here’s a breakdown of each phase:
+
+1. **earlyRead Phase**: 
+  This phase captures the current height of the `textarea` by reading the `offsetHeight` directly from the DOM. 
+  This read operation from the DOM is necessary because the textarea can also be resized manually by the user, so its size must be checked before any adjustment.
+  The result, `currentHeight`, is passed to the next phase. 
+  In this phase, we use the `extraHeight` as our tracked dependency to ensure that the code will run multiple times.
+  We encourage you to remove this statement: `console.log('earlyRead: extra height:', this.extraHeight());`.
+  If you do this, you will see that the `earlyRead` callback will only execute once and that any manual change to the textarea will be ignored in the next run.
+
+2. **write Phase**: 
+  The write phase adds the `extraHeight` value to the captured `currentHeight` and updates the textarea’s height style property.
+  This DOM write operation directly adjusts the element’s height in pixels.
+  An `onCleanup` function is provided to handle any required cleanup or resources before the next write operation.
+  In this example no cleanup is required, but we wanted to mention the fact that long-running tasks (such as a timeout) should be cleaned up. The cleanup will be called before entering the same phase again, or if the effect itself is destroyed.
+  The `write` phase then passes the new height, `newHeight`, to the `read` phase.
+  Hint: Pass the same value to `read` (eg. `return 100`) and you will see that the follow-up phase won't be executed.
+  Setting the same number twice won’t be considered a change, so the `write` phase won’t mark the `read` phase as dirty.
+
+3. **read Phase**: The final `read` phase logs the `newHeight`. 
+  We could also read from the DOM in that phase and store the result to a new signal. But in this example this work is not necessary, because the `earlyRead` is already doing that job.
```

Comment:
```suggestion
1. **`earlyRead` Phase**: 
  This phase captures the current height of the `textarea` by reading the `offsetHeight` directly from the DOM. 
  This read operation from the DOM is necessary because the textarea can also be resized manually by the user, so its size must be checked before any adjustment.
  The result, `currentHeight`, is passed to the next phase. 
  In this phase, we use the `extraHeight` as our tracked dependency to ensure that the code will run multiple times.
  We encourage you to remove this statement: `console.log('earlyRead: extra height:', this.extraHeight());`.
  If you do this, you will see that the `earlyRead` callback will only execute once and that any manual change to the textarea will be ignored in the next run.

2. **`write` Phase**: 
  The write phase adds the `extraHeight` value to the captured `currentHeight` and updates the textarea’s height style property.
  This DOM write operation directly adjusts the element’s height in pixels.
  An `onCleanup` function is provided to handle any required cleanup or resources before the next write operation.
  In this example no cleanup is required, but we wanted to mention the fact that long-running tasks (such as a timeout) should be cleaned up. The cleanup will be called before entering the same phase again, or if the effect itself is destroyed.
  The `write` phase then passes the new height, `newHeight`, to the `read` phase.
  Hint: Pass the same value to `read` (e.g. `return 100`) and you will see that the follow-up phase won't be executed.
  Setting the same number twice won’t be considered a change, so the `write` phase won’t mark the `read` phase as dirty.

3. **`read` Phase**: The final `read` phase logs the `newHeight`. 
  We could also read from the DOM in that phase and store the result to a new signal. But in this example this work is not necessary, because the `earlyRead` is already doing that job.
```